src/main/java/com/google/api/generator/engine/ast/TernaryExpr.java

```diff
@@ -29,10 +29,13 @@
   public TypeNode type() {
     TypeNode thenType = thenExpr().type();
     TypeNode elseType = elseExpr().type();
-    if (thenType.equals(elseType)
-        || (thenType.isSupertypeOrEquals(elseType) && !thenType.equals(TypeNode.NULL))) {
+    if (thenType.isSupertypeOrEquals(elseType) && !thenType.equals(TypeNode.NULL)) {
       return thenType;
     }
+    if (thenType.equals(elseType)) {
+      // If types are boxed/primitive type equals, return the boxed type.
+      return TypeNode.isReferenceType(thenType) ? thenType : elseType;
+    }
```

Comment:
add another condition based on the discussion, that when the types are boxed/primitive equal, we should the boxed type as the whole expression type.

Reply:
Let's clean this up a bit more by separating them into a primitive-only check (that includes the boxed equality checking) and a purely-reference type check.

Reply:
Sounds cool, first check `if (thenType.isPrimitiveType() || elseType.isPrimitiveType())` covers the condition either/both of types are primitive, then return the equal primitive type or the boxed type. Second check covers the reference types, we should always return the super /not-null type.